### PR TITLE
[WIP] Instrument services with tritium metrics and performance trace logging

### PR DESCRIPTION
**Pull request by schlosna**
**Goals (and why)**: Capture aggregate response time and service call metrics for key value, timestamp, and lock services.

**Implementation Description (bullets)**: Utilizes [tritium](https://github.com/palantir/tritium) to wrap the following services with metric and logging instrumentation:

 * `KeyValueService`
 * `TimestampService`
 * `RemoteLockService`
 * `PaxosLearner`
 * `PaxosAcceptor`
 * `PingableLeader`

Exposes cache effectiveness metrics such as hit ratio for `TimestampCache`.

**Concerns (what feedback would you like?)**: Initial WIP, need to include example output.

**Where should we start reviewing?**: `ServiceCreator#instrument`

**Priority (whenever / two weeks / yesterday)**: this week

Addresses #1365
Extends #689 
Related to #1454 
Supersedes #1625

### CODE REVIEW

**Comment by schlosna** [comment]
This is not yet ready for review, but wanted to get thoughts. @hsaraogi for SA.

**Comment by hsaraogi** [comment]
Overall looks good, did not read through in detail yet. Some concerns/questions:

1. Are we sure that the performance overhead for any of the operations are insignificant with the instrument wrapping? 
2. Benchmarks will catch the KVS level ones, worth running them against this branch.
3. Need not be a part of the PR, but since we have kvs and lockservice level metrics on the client side, can we delete the dropwizard timers/meters used in say `SnapshotTransaction` (this may break some existing dashboards), but are they giving us any additional info? eg: 
`Timer.Context timer = getTimer("getRows").time();` simply measure the time for the getRows kvs operation.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/cache/TimestampCache.java [hunk 1]
```diff
@@ -15,32 +15,43 @@
...
+
 import com.google.common.cache.Cache;
 import com.google.common.cache.CacheBuilder;
+import com.palantir.atlasdb.util.AtlasDbMetrics;
 
 /**
- * This class just here for readability and not directly leaking / tying us down to a Guava class in our API
+ * This class just here for readability and not directly leaking / tying us down to a Guava class in our API.
  */
 public class TimestampCache {
```
Comment (schlosna) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/cache/TimestampCache.java]: - [ ] add unit tests

File: atlasdb-client/src/test/java/com/palantir/atlasdb/util/AtlasDbMetricsTest.java [hunk 2]
```diff
@@ -44,4 +61,40 @@ public void metricsIsNotDefaultWhenSet() {
     public void nullMetricsCannotBeSet() {
         AtlasDbMetrics.setMetricRegistry(null);
     }
+
```
Comment (schlosna) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/util/AtlasDbMetricsTest.java]: - [ ] remove blank line

File: atlasdb-client/src/test/java/com/palantir/atlasdb/util/AtlasDbMetricsTest.java [hunk 3]
```diff
@@ -44,4 +61,40 @@ public void metricsIsNotDefaultWhenSet() {
...
+
+    @Test
+    public void instrumentWithDefaultName() throws Exception {
+        MetricRegistry metrics = MetricRegistries.createWithHdrHistogramReservoirs();
+        AtlasDbMetrics.setMetricRegistry(metrics);
```
Comment (schlosna) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/util/AtlasDbMetricsTest.java]: - [ ] save/restore metric in before/after, possibly create `@Rule` for metrics

File: atlasdb-config/src/test/java/com/palantir/atlasdb/factory/LeadersTest.java [hunk 4]
```diff
@@ -15,31 +15,61 @@
...
 import com.palantir.paxos.PaxosAcceptor;
 import com.palantir.paxos.PaxosLearner;
+import com.palantir.paxos.PaxosValue;
+import com.palantir.tritium.metrics.MetricRegistries;
 
 public class LeadersTest {
 
     public static final Set<String> REMOTE_SERVICE_ADDRESSES = ImmutableSet.of("foo:1234", "bar:5678");
 
+    private MetricRegistry beforeMetrics;
```
Comment (schlosna) [inline atlasdb-config/src/test/java/com/palantir/atlasdb/factory/LeadersTest.java]: - [ ] create  & use `@Rule` for metrics

File: atlasdb-dbkvs/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/TableValueStyleCacheTest.java [hunk 5]
```diff
@@ -63,6 +65,7 @@ public void setup() {
 
     @After
     public void tearDown() {
+        ConsoleReporter.forRegistry(AtlasDbMetrics.getMetricRegistry()).build().report();
```
Comment (schlosna) [inline atlasdb-dbkvs/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/TableValueStyleCacheTest.java]: - [ ] use `@Rule` for metrics
- [ ] close reporter

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/AtlasDbTestCase.java [hunk 6]
```diff
@@ -82,16 +91,30 @@ public boolean isStandaloneServer() {
         }
     }
 
+    @BeforeClass
+    public static void savePreviousMetrics() throws Exception {
+        previousMetrics = AtlasDbMetrics.getMetricRegistry();
```
Comment (schlosna) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/AtlasDbTestCase.java]: - [ ] move to `@Rule`

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/keyvalue/impl/InstrumentedKeyValueServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    @Override
+    protected KeyValueService getKeyValueService() {
+        return AtlasDbMetrics.instrument(KeyValueService.class,
+                new InMemoryKeyValueService(false),
+                METRIC_PREFIX);
+    }
+
+    @BeforeClass
+    public static void beforeClass() throws Exception {
+        previousMetrics = AtlasDbMetrics.getMetricRegistry();
```
Comment (schlosna) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/keyvalue/impl/InstrumentedKeyValueServiceTest.java]: - [ ] Use `@Rule`

File: timelock-server/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosSynchronizer.java [hunk 8]
```diff
@@ -58,15 +58,21 @@ public static void synchronizeLearner(PaxosLearner learnerToSynchronize,
...
-                .filter(response -> response.paxosValue() != null)
-                .map(PaxosValueResponse::paxosValue)
-                .max(Comparator.comparingLong(PaxosValue::getRound));
+        try {
+            List<PaxosValueResponse> responses = PaxosQuorumChecker.collectAsManyResponsesAsPossible(
+                    ImmutableList.copyOf(paxosLearners),
+                    learner -> ImmutablePaxosValueResponse.of(learner.getGreatestLearnedValue()),
+                    executor,
+                    PaxosQuorumChecker.DEFAULT_REMOTE_REQUESTS_TIMEOUT_IN_SECONDS);
+            executor.shutdown();
```
Comment (schlosna) [inline timelock-server/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosSynchronizer.java]: - [ ] split to separate PR

**Comment by schlosna** [comment]
@hsaraogi Here are some JMH microbenchmarks for Tritium instrumentation overhead, which currently adds about ~500 nanoseconds per invocation for the current metrics & trace logging instrumentation. The assumption is that the majority of operations being instrumented are significantly more expensive (typically hitting remote system), so this overhead should be negligible in the grand scheme of things and more than pays for itself.

```
Benchmark                                         Mode  Cnt    Score    Error  Units
ProxyBenchmark.raw                                avgt    5    3.129 ±  0.434  ns/op
ProxyBenchmark.instrumentedWithoutHandlers        avgt    5    3.174 ±  0.278  ns/op
ProxyBenchmark.instrumentedWithLogging            avgt    5   89.622 ±  5.823  ns/op
ProxyBenchmark.instrumentedWithMetrics            avgt    5  362.495 ± 24.333  ns/op
ProxyBenchmark.instrumentedWithMetricsAndLogging  avgt    5  507.094 ± 47.811  ns/op
ProxyBenchmark.instrumentedWithTracing            avgt    5  766.690 ± 14.486  ns/op
```

File: atlasdb-client/src/test/java/com/palantir/atlasdb/util/MetricsRule.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+import com.codahale.metrics.ConsoleReporter;
+import com.codahale.metrics.MetricRegistry;
+
+public class MetricsRule extends ExternalResource {
+
+    private MetricRegistry previousMetrics;
+
+    @Override
+    protected void before() throws Throwable {
```
Comment (hsaraogi) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/util/MetricsRule.java]: Maybe a stupid question, but I am very confused as ``previousMetrics`` and ``metrics`` will be the same thing unless someone else ``sets``(only one call to set is made, which is in this class), noone else does and why are we reporting from the ``newMetrics`` after we reset the registry on line 37.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/AtlasDbTestCase.java [hunk 10]
```diff
@@ -113,12 +125,20 @@ public void setUp() throws Exception {
         txManager = new CachingTestTransactionManager(txManager);
     }
 
+    @After
+    public void after() throws Exception {
```
Comment (himangi774) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/AtlasDbTestCase.java]: Isnt this a part of the ``MetricsRule`` already?

File: build.gradle [hunk 11]
```diff
@@ -70,6 +70,10 @@ allprojects {
         propertiesFile file: project.rootProject.file('versions.props')
     }
 
+    configurations {
+        testArtifacts.extendsFrom testCompile
```
Comment (himangi774) [inline build.gradle]: Can we add this to only required modules?

File: atlasdb-client/src/main/java/com/palantir/atlasdb/util/AtlasDbMetrics.java [hunk 12]
```diff
@@ -39,16 +45,38 @@ public static synchronized void setMetricRegistry(MetricRegistry metricRegistry)
...
+        MetricRegistry currentMetricRegistry = AtlasDbMetrics.metrics;
+        if (currentMetricRegistry == null) {
             synchronized (AtlasDbMetrics.class) {
-                if (AtlasDbMetrics.metrics == null) {
-                    AtlasDbMetrics.metrics = SharedMetricRegistries.getOrCreate(DEFAULT_REGISTRY_NAME);
-                    log.info("Metric Registry was not set, setting to default registry name of "
+                currentMetricRegistry = AtlasDbMetrics.metrics;
+                if (currentMetricRegistry == null) {
+                    currentMetricRegistry = SharedMetricRegistries.getOrCreate(DEFAULT_REGISTRY_NAME);
+                    log.warn("Metric Registry was not set, setting to shared default registry name of "
```
Comment (schlosna) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/util/AtlasDbMetrics.java]: My thought here is that folks really want to inject their own AtlasDB metric registry before anything attempts to use is it, because at that point it is likely to late to replace the global one, and they won't capture stats & instrumentation.

File: atlasdb-client/src/test/java/com/palantir/atlasdb/util/MetricsRule.java (hunk 9 above)
Comment (schlosna) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/util/MetricsRule.java]: The intent with the `MetricsRule` is that due to the nastiness around `AtlasDbMetrics` holding on to a shared global `MetricRegistry`, we want to isolate our tests to be able to assert and verify the metrics captured for an individual test case (using `@Rule public MetricsRule ...`) or entire test class (using `@ClassRule public static MetricsRule ...`).

File: atlasdb-client/src/main/java/com/palantir/atlasdb/util/AtlasDbMetrics.java (hunk 12 above)
Comment (himangi774) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/util/AtlasDbMetrics.java]: and why do we have the new variable `currentMetricRegistry`?

File: atlasdb-dbkvs/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/TableValueStyleCacheTest.java [hunk 13]
```diff
@@ -45,6 +47,9 @@
     private final ConnectionSupplier connectionSupplier = mock(ConnectionSupplier.class);
     private final TableValueStyleCache valueStyleCache = new TableValueStyleCache();
 
+    @Rule
+    public MetricsRule metricsRule = new MetricsRule();
```
Comment (himangi774) [inline atlasdb-dbkvs/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/TableValueStyleCacheTest.java]: This one is not instrumented, right? We dont need the ``MetricsRule`` here or we can check that no metrics are registered in the tests.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/util/AtlasDbMetrics.java [hunk 14]
```diff
@@ -1,12 +1,12 @@
 /**
  * Copyright 2017 Palantir Technologies
- *
+ * <p>
```
Comment (schlosna) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/util/AtlasDbMetrics.java]: Yeah, this is a side effect of the IntelliJ auto-formatter because AtlasDB checkstyle setup is incorrect (see #1693 )

File: atlasdb-client/src/main/java/com/palantir/atlasdb/util/AtlasDbMetrics.java (hunk 12 above)
Comment (schlosna) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/util/AtlasDbMetrics.java]: The intent with `currentMetricRegistry` is to minimize the happy read path to a single volatile read, where previously we would do at least 2 volatile reads when the registry was already set.

File: atlasdb-client/src/test/java/com/palantir/atlasdb/util/MetricsRule.java (hunk 9 above)
Comment (hsaraogi) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/util/MetricsRule.java]: But why is the console output necessary, we are not verifying it.

Comment (schlosna) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/util/MetricsRule.java]: The console output isn't necessary, it is for awareness what operations that test performed.

**Comment by schlosna** [comment]
@hsaraogi I think I addressed all your comments.

**Comment by hsaraogi** [comment]
Main blocker:

Running benchmarks against this PR shows a significant increase in the time of operations. 
```
Benchmark                                                            DB                            Percentage change
KvsGetRangeBenchmarks.getSingleRangeDirty                            ORACLE                        25
KvsGetRangeBenchmarks.getSingleRange                                 CASSANDRA                     25
TransactionGetDynamicBenchmarks.getFirstColumnExplicitlyGet          CASSANDRA                     27
TransactionGetDynamicBenchmarks.getAllColumnsImplicitly              ORACLE                        28
KvsGetDynamicBenchmarks.getFirstColumnExplicitly                     CASSANDRA                     28
KvsGetRangeBenchmarks.getSingleRangeDirty                            CASSANDRA                     28
SweepBenchmarks.multipleUniformSweepRun                              POSTGRES                      29
TransactionGetDynamicBenchmarks.getAllColumnsImplicitly              POSTGRES                      33
KvsGetRowsColumnRangeBenchmarks.getAllColumnsAligned                 CASSANDRA                     33
TransactionGetBenchmarks.getRangeDirty                               CASSANDRA                     62
```

**Comment by hsaraogi** [comment]
Benchmarks after revision are as follows.
```
Benchmark                                                            DB                            Percentage change
KvsGetRangeBenchmarks.getSingleRangeDirty                            ORACLE                        14
KvsGetRangeBenchmarks.getSingleRange                                 CASSANDRA                     15
TransactionGetDynamicBenchmarks.getFirstColumnExplicitlyGet          CASSANDRA                     11
TransactionGetDynamicBenchmarks.getAllColumnsImplicitly              ORACLE                        6
KvsGetDynamicBenchmarks.getFirstColumnExplicitly                     CASSANDRA                     12
KvsGetRangeBenchmarks.getSingleRangeDirty                            CASSANDRA                     13
SweepBenchmarks.multipleUniformSweepRun                              POSTGRES                      11
TransactionGetDynamicBenchmarks.getAllColumnsImplicitly              POSTGRES                      34
KvsGetRowsColumnRangeBenchmarks.getAllColumnsAligned                 CASSANDRA                     19
TransactionGetBenchmarks.getRangeDirty                               CASSANDRA                     -28
```
